test/Constraints/diagnostics.swift

```diff
@@ -1572,3 +1572,14 @@ func testNilCoalescingOperatorRemoveFix() {
   if ("" // This is a comment
       ?? "").isEmpty {} // expected-warning {{left side of nil coalescing operator '??' has non-optional type 'String', so the right side is never used}} {{1572:9-1573:12=}}
 }
+
+// https://github.com/apple/swift/issues/73029
+func testBinaryOpWrongTypesFix() {
+  struct S<T> {
+    static func ==(lhs: S, rhs: S) -> Bool {
+      true
+    }
+  }
+  func getS<T>(value: T) -> S<T> {}
+  let _ = getS(value: "") == getS(value: 0) // expected-error {{binary operator '==' cannot be applied to operands of type 'S<String>' and 'S<Int>'}}
```

Comment:
I think we need to make a couple of changes to properly fix this example: 

1. We need to make sure that all of the unrelated overloads of `==` don't interfere here, currently there are 4 solutions (including two with missing conformances on `BinaryInteger` and `_Pointer`) when we should have only 2 that are related to `==(_: S, _: S) -> Bool` overload. This could be done by adjusting the score of `MissingConformance` fix which is currently `1` and is too low (this might cause a chain reaction though because argument-to-parameter conversion score is 2).

2. Once we are at two solutions with incorrect arguments it would be possible to produce a conversion mismatch diagnostic with a note that points out that `String` and `Int` are expected to be equal just like this case with no overloads:
```
struct S<T> {
}

func test<T>(_: S<T>, _: S<T>) {}

test(S<Int>(), S<String>())
```

Reply:
See latest here, with the effects of increasing `MissingConformance` from `1` to `2`.

Improves diagnoses for:
```
Generics/conditional_conformances_literals.swift
Constraints/diag_ambiguities.swift
Constraints/rdar85263844_swift6.swift
type/implicit_some/opaque_parameters.swift
type/builtin_types.swift
type/opaque_parameters.swift
stdlib/UnicodeScalarDiagnostics.swift
expr/expressions.swift
decl/protocol/existential_member_accesses_self_assoctype.swift
```
Regression of diagnoses for:
```
Generics/inheritance.swift
Generics/inverse_copyable_requirement.swift
Macros/macro_expand.swift
type/subclass_composition_objc.swift
StringProcessing/Parse/forward-slash-regex.swift
```
Changes but not really better or worse:
```
Constraints/diagnostics.swift
Constraints/rdar40002266.swift
Constraints/operator.swift
Parse/confusables.swift
type/subclass_composition.swift
```
@xedin Reasonable to continue in this direction despite the regressions?